cider/tests/utils/CiderBatchStringifier.cpp

```diff
@@ -92,6 +95,51 @@ std::string StructBatchStringifier::stringifyValueAt(CiderBatch* batch, int row_
   return row.getString();
 }
 
+uint8_t DecimalBatchStringifier::getScale(const ScalarBatch<__int128_t>* batch) {
+  auto type_str = std::string(batch->getArrowFormatString());
+  uint8_t scale = std::stoi(type_str.substr(type_str.find(',') + 1));
+  return scale;
+}
+
+uint8_t DecimalBatchStringifier::getWidth(const ScalarBatch<__int128_t>* batch) {
+  auto type_str = std::string(batch->getArrowFormatString());
+  auto start = type_str.find(':') + 1;
+  auto end = type_str.find(',');
+  uint8_t width = std::stoi(type_str.substr(start, end - start));
+  return width;
+}
+
+std::string DecimalBatchStringifier::stringifyValueAt(CiderBatch* batch, int row_index) {
+  auto scalar_batch = batch->as<ScalarBatch<__int128_t>>();
+  if (!scalar_batch) {
+    CIDER_THROW(CiderRuntimeException,
+                "ScalarBatch is nullptr, maybe check your casting?");
+  }
+
+  auto scale = getScale(scalar_batch);
+  auto width = getWidth(scalar_batch);
+  auto data_buffer = scalar_batch->getRawData();
+  auto valid_bitmap = scalar_batch->getNulls();
+
+  if (valid_bitmap && !CiderBitUtils::isBitSetAt(valid_bitmap, row_index)) {
+    return NULL_VALUE;
+  } else {
+    __int128_t value = data_buffer[row_index];
+
+    if (!scale) {
+      // integral type
+      return CiderInt128Utils::Int128ToString(value);
```

Comment:
I am wondering for this `scale==0` case, it also can be covered by `CiderInt128Utils::Decimal128ToDouble(value, width, scale)`.

Reply:
The stringifier treats integral types and floating point types differently. Integral types are directly stringified with `to_string`, while floating points are stringified with 16 significant digits.

`scale == 0` means the value being stringified here is an `int128` instance. Its value is expected to be checked with another reference integer (which typically comes from Cider), and in this case we need not (and should not) set any significant digit. Otherwise, the value is a decimal, which will be casted to double and compare with float/double values.